languages/javalang/builder-renderer/commercetools-java-client/src/main/java/client/ApiHttpClient.java

```diff
@@ -1,9 +1,45 @@
 package client;
 
-import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
 
-public interface ApiHttpClient {
+public class ApiHttpClient {
 
-    public ApiHttpResponse execute(final ApiHttpRequest httpRequest) throws IOException;
-    
+    private final Middleware reducedMiddleware;
+
+    public ApiHttpClient(final List<Middleware> middlewares) {
+        final List<Middleware> middlewareList = new ArrayList<>();
+        middlewareList.addAll(middlewares);
+        middlewareList.add(NoOpMiddleware.getInstance());
+        this.reducedMiddleware = middlewareList.stream().reduce(ApiHttpClient::reduceMiddlewares).get();
+    }
+
+    public CompletableFuture<ApiHttpResponse> execute(final ApiHttpRequest request) {
+        return reducedMiddleware
+                .next(MiddlewareArg.from(request, null, null, null))
```

Comment:
This `from()`call with all the `null`looks a bit unusual. May be it would be better to offer a constructor just accepting a `request`? WDYT?

Reply:
Sounds like a good idea, at the start we always have only the request.